File: 334.递增的三元子序列.go
/*
 * @lc app=leetcode.cn id=334 lang=golang
 *
 * [334] 递增的三元子序列
 */

// @lc code=start
func increasingTriplet(nums []int) bool {
	length := len(nums)
	if length < 3 {
		return false
	}
	minSlice := make([]int, len(nums))
	maxSlice := make([]int, len(nums))
	for i, j := 0, length-1; i < length && j >= 0; i, j = i+1, j-1 {
		if i == 0 {
			minSlice[0] = nums[0]
		} else {
			minSlice[i] = min(nums[i], minSlice[i-1])
		}
		if j == length-1 {
			maxSlice[j] = nums[j]
		} else {
			maxSlice[j] = max(nums[j], maxSlice[j+1])
		}
	}
	for j := 1; j < length-1; j++ {
		if nums[j] > minSlice[j] && nums[j] < maxSlice[j] {
			return true
		}
	}

	// 三循环，限制后者大于前者，小于等于皆不可以
	// for i := 0; i < length-2; i++ {
	//     for j := i+1; j < length-1; j++ {
	//         if nums[j] <= nums[i] {
	//             continue
	//         }
	//         for h := j+1; h < length; h++ {
	//             if nums[h] <= nums[j] {
	//                 continue
	//             }
	//             return true
	//         }
	//     }
	// }
	return false
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// @lc code=end

File: 334.递增的三元子序列.py
#
# @lc app=leetcode.cn id=334 lang=python3
#
# [334] 递增的三元子序列
#

# @lc code=start
class Solution:
    def increasingTriplet(self, nums: List[int]) -> bool:
        length = len(nums)
        if length < 3:
            return False
        min_list = [0 for _ in range(length)]
        max_list = [0 for _ in range(length)]
        i, j = 0, length-1
        while i < length and j >= 0:
            if i == 0:
                min_list[i] = nums[0]
            else:
                min_list[i] = min(nums[i], min_list[i-1])
            i += 1
            if j == length - 1:
                max_list[j] = nums[j]
            else:
                max_list[j] = max(nums[j], max_list[j+1])
            j -= 1
        print(min_list)
        print(max_list)
        for i in range(1, length):
            if nums[i] > min_list[i] and nums[i] < max_list[i]:
                return True
        return False
# @lc code=end
